rfcs/0003-multi-tenancy-mode/README.md

```diff
@@ -0,0 +1,197 @@
+```
+
+When running in the `multi-tenant` mode, Flux behaves differently:
+
+- The source-to-cluster reconciliation no longer runs under the service account of
+  the Flux controllers. The controller service account, is only used to impersonate
+  the service account specified in the Flux custom resources (`Kustomizations`, `HelmReleases`).
+- When no service account name is specified in a Flux custom resource,
+  a default will be used e.g. `system:serviceaccount:<tenant-namespace>:flux`.
+- When a Flux custom resource (`Kustomizations`, `HelmReleases`, `ImagePolicies`, `ImageUpdateAutomations`)
+  refers to a source in a different namespace, access is granted based the source access control list.
+  If no ACL is defined for a source, cross-namespace access is denied.
+- When a Flux notification (`Alerts`, `Receivers`)
+  refers to a resource in a different namespace, access is denied.
+
+### Tenants Onboarding
+
+When onboarding tenants, platform admins should have the option to assign namespaces, set
+permissions and register the tenants repositories onto clusters in a declarative manner. 
+
+The Flux CLI offers an easy way of generating all the Kubernetes manifests needed to onboard tenants:
+
+- `flux create tenant` command generates namespaces, service accounts and Kubernetes RBAC
+  with restricted access to the cluster resources, given tenants access only to their namespaces.
+- `flux create secret git` command generates SSH keys used by Flux to clone the tenants repositories.
+- `flux create source git` command generates the configuration that tells Flux which repositories belong to tenants.
+- `flux create kustomization` command generates the configuration that tells Flux how to reconcile the manifests found in the tenants repositories.
+
+All the above commands have an `--export` flag for generating the Kubernetes resources in YAML format.
+The platform admins should place the generated manifests in the repository that defines the cluster(s) desired state.
+
+Here is an example of the generated manifests:
+
+```yaml
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: tenant1
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: flux
+  namespace: tenant1
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: flux
+  namespace: tenant1
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: cluster-admin
+subjects:
+  - kind: ServiceAccount
+    name: flux
+    namespace: tenant1
+---
+apiVersion: source.toolkit.fluxcd.io/v1beta1
+kind: GitRepository
+metadata:
+  name: tenant1
+  namespace: tenant1
+spec:
+  interval: 5m0s
+  ref:
+    branch: main
+  secretRef:
+    name: tenant1-git-auth
+  url: ssh://git@github.com/org/tenant1
+---
+apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
+kind: Kustomization
+metadata:
+  name: tenant1
+  namespace: tenant1
+spec:
+  interval: 10m0s
+  path: ./
+  prune: true
+  serviceAccountName: flux
```

Comment:
With this proposed method, the kustomize-controller or helm-controller can reconcile _any_ service account within the namespace of the CR. With this in mind, a user needs to be hyper-aware that any service account in that namespace could be used for reconciliation and must not role-bind to give any service account any kind of elevated privilege that it did not intend for the flux reconcilers (such as giving access to create HelmReleases across the cluster, which essentially opens up Pandora's box). Will this be noted somewhere specifically/is there a way to ensure that flux can only reconcile with service accounts with .metadata.labels intended specifically for flux?

Reply:
I think we shouldn't enforce any limitations for the service accounts that are provisioned by platform admins because they may chose to allow a tenant to own multiple namespaces, so they can bind a SA to all those namespaces. A tenant can't create cluster role bindings so it's not possible for them to evade their boundary.

Reply:
Wouldn't this same concern apply to e.g. a `Deployment` which accepts any arbitrary service account as a configuration value? Which I assume, isn't a problem. Raising the question why it would be different for a Flux Custom Resource?

Reply:
> Wouldn't this same concern apply to e.g. a `Deployment` which accepts any arbitrary service account as a configuration value?

These are the same technically -- a Deployment can run a container that uses kubectl to cause mischief. But the opportunity for mistakes is higher when Flux is introduced. With _only_ Deployments and other "normal" objects you would not usually consider creating a service account with wide-ranging permissions. When you're using Flux, you have much more (RBAC) to think about.

Reply:
> With only Deployments and other "normal" objects you would not usually consider creating a service account with wide-ranging permissions. 

Depend on what you consider to be "normal" here, every single Kubernetes addon needs wide-ranging permissions, are these abnormal?

Reply:
Yes, those are abnormal, in this context. The equivalence Hidde was inviting us to draw was between a Deployment and (e.g.,) a HelmRelease, as used by a tenant. They both have a field for supplying a service account, so why bother making especial mention of it for HelmReleases?

My point is: if I'm an admin setting permissions up for a tenant to be able to run a Deployment, all it really needs to be able to do is run pods and talk over the network to other processes. But if I'm setting permissions up for a tenant to be able to create a HelmRelease, the risk is much higher, because it will need to have permissions to create other resources. As Jonathan says, you need to be hyper-aware that such a service account will be usable by anything in that namespace.

Reply:
I think my point is more that if you make extensive use of RBAC, you should be aware of what this entails (for your tenants). 

The difference between not sharing a service account between multiple resources, or using any arbitrary account that is available to you, does not change for a `Deployment` or a Flux resource.

Reply:
Not saying here that this means we should not highlight or be upfront about it, but rather that I am not really enthusiastic about adding e.g. label selectors for service account.

Reply:
> I am not really enthusiastic about adding e.g. label selectors for service accounts.

Right -- requiring labels is not a protection, since any process that can run with the service account can do mischief, not just the Flux controllers.

Reply:
The main challenge on the model in which the tenant's service account is created in the tenant's namespace is that anyone with `pod create` permissions in the tenant's namespace would be able to privesc to whatever permissions the flux tenant service account is running as - current example is cluster-admin within that namespace.

In this model platform admins may have to lean on admission controllers to block that path to privilege escalation.

Reply:
I added some recommendations on how the tenant can be structured to protect their service accounts in the RFC001 (xref: https://github.com/fluxcd/flux2/pull/2086#discussion_r767691190) without requiring admission controllers.